patches/minecraft/net/minecraft/entity/passive/EntityVillager.java.patch

```diff
@@ -17,23 +17,189 @@
++        p_70014_1_.func_74778_a("professionName", professionName);
+         p_70014_1_.func_74768_a("Riches", this.field_70956_bz);
+         p_70014_1_.func_74768_a("Career", this.field_175563_bv);
+         p_70014_1_.func_74768_a("CareerLevel", this.field_175562_bw);
+@@ -283,14 +286,34 @@
+     public void func_70037_a(NBTTagCompound p_70037_1_)
+     {
+         super.func_70037_a(p_70037_1_);
+-        this.func_70938_b(p_70037_1_.func_74762_e("Profession"));
++        int professionId = p_70037_1_.func_74762_e("Profession");
++        String professionName = p_70037_1_.func_74779_i("professionName");
++        net.minecraftforge.fml.common.registry.VillagerRegistry.VillagerProfession profession = net.minecraftforge.fml.common.registry.VillagerRegistry.getProfessionForName(professionName);
++        String oldProfession = profession.getName();
++        if(oldProfession.equals(professionName) && professionId != profession.getId())
++        {
++            professionId = profession.getId();
++        }
++        else if(!professionName.isEmpty() && !oldProfession.equals(professionName))
++        {
++            professionId = 0;
++        }
++        this.func_70938_b(professionId);
+         this.field_70956_bz = p_70037_1_.func_74762_e("Riches");
+         this.field_175563_bv = p_70037_1_.func_74762_e("Career");
+         this.field_175562_bw = p_70037_1_.func_74762_e("CareerLevel");
+         this.field_175565_bs = p_70037_1_.func_74767_n("Willing");
+-
+-        if (p_70037_1_.func_150297_b("Offers", 10))
++        
++        String newProfessionName = net.minecraftforge.fml.common.registry.VillagerRegistry.getProfessionForId(func_70946_n()).getName();
++        if(professionId != func_70946_n()||!(professionName.equals(newProfessionName)||professionName.isEmpty()))
+         {
++            this.field_175563_bv = 0;
++            this.field_175562_bw = 0;
++            this.field_70963_i = null;
++            this.func_70938_b(func_70946_n());
++        }
++        else if (p_70037_1_.func_150297_b("Offers", 10))
++        {
+             NBTTagCompound nbttagcompound1 = p_70037_1_.func_74775_l("Offers");
+             this.field_70963_i = new MerchantRecipeList(nbttagcompound1);
+         }
+@@ -333,14 +356,27 @@
  
-     private void func_175554_cu()
+     public void func_70938_b(int p_70938_1_)
+     {
++        this.field_175563_bv = 0;
++        this.field_175562_bw = 0;
++        this.field_70963_i = null;
+         this.field_70180_af.func_75692_b(16, Integer.valueOf(p_70938_1_));
+     }
+ 
+     public int func_70946_n()
+     {
+-        return Math.max(this.field_70180_af.func_75679_c(16) % 5, 0);
++        return Math.max(this.field_70180_af.func_75679_c(16) % net.minecraftforge.fml.common.registry.VillagerRegistry.VillagerProfession.nextId, 0);
+     }
+ 
++    public void setCareer(int careerId)
++    {
```

Comment:
I will move it to end of file.  for validating, it would require essentially a modification of the populateBuyLists method inside the file (as that only checks if both career and level are zero), I can add that though.